CONTRIBUTING.md

```diff
@@ -0,0 +1,113 @@
+
+### New features
+
+If you want to add a new major feature, please reach out to us via [Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA) to discuss the specifics. By doing so, we can help guide you and ensure that it's feasible before you spend a considerable amount of time.
+
+## Building & developing
+
+We recommend that you use [IntelliJ](https://www.jetbrains.com/idea/) for development, as some of the Gradle specifics in `openrewrite/rewrite` are not supported as well in other IDEs. 
+
+All projects use Java 17 for their tests, to take advantage of text blocks. For `src/main`, we use Java 8 exclusively so that our users only need Java 8 to run OpenRewrite migration recipes. We use [Gradle](https://gradle.org/) as our build tool.
+
+**Note**: The Gradle wrapper requires a preexisting JDK 8 installation. If not present, the build will fail to download the toolchain in the `:write-gradle:compilyGroovy` task. You will need to install JDK 8 manually. For Unix systems, we strongly commend using [sdkman](https://sdkman.io/) to install and switch between Java versions. 
+
+### IntelliJ changes
+
+By default, IntelliJ uses Gradle to build and run tests with. While that ensures compatibility, it is very slow. To help speed up compilation and testing, we recommend that you change this to use `IntelliJ IDEA` instead:
+
+![](/resources/gradle-select.png)
+
+As part of doing that, you'll also need to update the `Java Compiler` to set the `-parameters` compiler flag. If your system does not have `UTF-8` as its default character encoding (e.g., Windows), you must also add `-encoding utf8`.
+
+![](/resources/parameters.png)
+
+You will also need to add an override to the compiler parameters for the `rewrite.rewrite-java-17.main` module to have the compilation options of: 
+
+```
+-parameters --add-exports jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED --add-exports jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED
+```
+
+![](/resources/compilation-options.png)
+
+#### Optimizing your IDE for only modules you want to work on
+
+If you are only working on a subset of the modules in this project, you can optimize your IDE to only load those modules. Create a new `IDE.properties` file if one doesn't exist in the base directory of the project and then copy the [IDE.properties.tmp](https://github.com/openrewrite/rewrite/blob/main/IDE.properties.tmp) file to it. Next, comment out any lines that correspond to modules that you do not want to work on. This will cause Gradle to swap those project dependencies for binary dependencies resolved from either Maven local or the OSS snapshots repository – which will speed up your IDE.
```

Comment:
```suggestion
If you are only working on a subset of the modules in [openrewrite/rewrite](https://github.com/openrewrite/rewrite), you can optimize your IDE to only load those modules. Create a new `IDE.properties` file if one doesn't exist in the base directory of the project and then copy the contents of [IDE.properties.tmp](https://github.com/openrewrite/rewrite/blob/main/IDE.properties.tmp) file to it. Next, comment out any lines that correspond to modules that you do not want to work on. This will cause Gradle to swap those project dependencies for binary dependencies resolved from either Maven local or the OSS snapshots repository – which will speed up your IDE.
```